h2integrate/controllers/openloop_controllers.py

```diff
@@ -125,7 +127,209 @@ def setup_partials(self):
+            f"{resource_name}_missed_load",
+            copy_shape=f"{resource_name}_in",
+            units=self.config.resource_units,
+            desc=f"{resource_name} missed load timeseries",
         )
+
+    def compute(self, inputs, outputs):
+        """
+        Compute the state of charge (SOC) and output flow based on demand and storage constraints.
+
+        """
+        resource_name = self.config.resource_name
+        max_capacity = self.config.max_capacity
+        max_charge_percent = self.config.max_charge_percent
+        min_charge_percent = self.config.min_charge_percent
+        init_charge_percent = self.config.init_charge_percent
+        max_charge_rate = self.config.max_charge_rate
+        max_discharge_rate = self.config.max_discharge_rate
+        charge_efficiency = self.config.charge_efficiency
+        discharge_efficiency = self.config.discharge_efficiency
+
+        # Initialize time-step state of charge prior to loop so the loop starts with
+        # the previous time step's value
+        soc = deepcopy(init_charge_percent)
+
+        demand_profile = inputs[f"{resource_name}_demand_profile"]
+
+        # initialize outputs
+        soc_array = outputs[f"{resource_name}_soc"]
+        curtailment_array = outputs[f"{resource_name}_curtailed"]
+        output_array = outputs[f"{resource_name}_out"]
+        missed_load_array = outputs[f"{resource_name}_missed_load"]
+
+        # Loop through each time step
+        for t, demand_t in enumerate(demand_profile):
+            # Get the input flow at the current time step
+            input_flow = inputs[f"{resource_name}_in"][t]
+
+            # Calculate the available charge/discharge capacity
+            available_charge = (max_charge_percent - soc) * max_capacity
+            available_discharge = (soc - min_charge_percent) * max_capacity
+
+            # Initialize persistent variables for curtailment and missed load
+            excess_input = 0.0
+            charge = 0.0
+
+            # Determine the output flow based on demand_t and SOC
+            if demand_t > input_flow:
+                # Discharge storage to meet demand
+                discharge_needed = (demand_t - input_flow) / discharge_efficiency
+                discharge = min(discharge_needed, available_discharge, max_discharge_rate)
+                soc -= discharge / max_capacity  # soc is a ratio with value between 0 and 1
+                output_array[t] = input_flow + discharge * discharge_efficiency
+            else:
+                # Charge storage with excess input
+                excess_input = (input_flow - demand_t) * charge_efficiency
+                charge = min(excess_input, available_charge, max_charge_rate)
+                soc += charge / max_capacity  # soc is a ratio with value between 0 and 1
+                output_array[t] = demand_t
+
+            # Ensure SOC stays within bounds
+            soc = max(min_charge_percent, min(max_charge_percent, soc))
```

Comment:
Is this already enforced from the `min` statements in the if/else cases above?

Reply:
Good to have an explicit check, as long as it doesn't alter the behavior in a negative way.